FirebaseMLModelDownloader/Sources/ModelDownloadManager.swift

```diff
@@ -0,0 +1,143 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import FirebaseCore
+
+enum DownloadStatus {
+  case notStarted
+  case inProgress
+  case completed
+}
+
+/// Progress and completion handlers for a model download.
+class DownloadHandlers {
+  typealias ProgressHandler = (Float) -> Void
+  typealias Completion = (Result<CustomModel, DownloadError>) -> Void
+
+  var progressHandler: ProgressHandler?
+  var completion: Completion
+
+  init(progressHandler: ProgressHandler?, completion: @escaping Completion) {
+    self.progressHandler = progressHandler
+    self.completion = completion
+  }
+}
+
+/// Manager for model downloads.
+class ModelDownloadManager: NSObject {
+  let app: FirebaseApp
+  var modelInfo: ModelInfo
+  var taskHandlers: [URLSessionDownloadTask: DownloadHandlers] = [:]
+
+  private(set) var downloadStatus: DownloadStatus = .notStarted
+
+  private lazy var downloadSession = URLSession(configuration: .ephemeral,
+                                                delegate: self,
+                                                delegateQueue: nil)
+
+  init(app: FirebaseApp, modelInfo: ModelInfo) {
+    self.app = app
+    self.modelInfo = modelInfo
+  }
+
+  /// Associate progress and completion handlers with download task.
+  private func setHandlers(for downloadTask: URLSessionDownloadTask, handlers: DownloadHandlers) {
+    taskHandlers[downloadTask] = handlers
+  }
+
+  /// Get download handlers for the given download task.
+  private func getHandlers(for downloadTask: URLSessionDownloadTask) -> DownloadHandlers? {
+    return taskHandlers[downloadTask]
+  }
+
+  /// Asynchronously download model file to device.
+  func startModelDownload(url: URL, progressHandler: DownloadHandlers.ProgressHandler? = nil,
+                          completion: @escaping DownloadHandlers.Completion) {
+    let downloadTask = downloadSession.downloadTask(with: url)
+    let downloadHandlers = DownloadHandlers(
+      progressHandler: progressHandler,
+      completion: completion
+    )
+    setHandlers(for: downloadTask, handlers: downloadHandlers)
+    downloadTask.resume()
+    downloadStatus = .inProgress
+  }
+}
+
+/// Extension to handle delegate methods.
+extension ModelDownloadManager: URLSessionDownloadDelegate {
+  func urlSession(_ session: URLSession,
+                  downloadTask: URLSessionDownloadTask,
+                  didFinishDownloadingTo location: URL) {
+    guard let handlers = getHandlers(for: downloadTask) else { return }
+    downloadStatus = .completed
+    let savedURL = ModelFileManager.modelsDirectory
+      .appendingPathComponent(downloadedModelFileName)
+    do {
+      try ModelFileManager.moveFile(at: location, to: savedURL)
```

Comment:
should we check if there is an existing file at the location before the move attempt?